### PR TITLE
First structure and working functions

### PR DESCRIPTION
**Pull request by Perados**
This adds a first Emarsys class which can make sync and async authenticated calls to the Emarsys API.
Once this pr will be merged, the rest of the classes and implemented endpoints will follow the same structure and logic. This project will most likely remain in pre-alpha state until all the endpoints are implemented and tested. Once that is done, it will be passed to beta state until everything is stable.

Also, once the structure is defined, documentation will be added.

### CODE REVIEW

File: pymarsys/emarsys.py
Comment (misterjoa) [inline pymarsys/emarsys.py]: this could be an init parameter with default value as is 'uri'

File: pymarsys/emarsys.py
Comment (misterjoa) [inline pymarsys/emarsys.py]: i don't know the best practice here, but what if this async code is used in another loop ?
```
new_event_loop()
```

File: pymarsys/emarsys.py
Comment (misterjoa) [inline pymarsys/emarsys.py]: repeated code. maybe factor it in a function
```
check_http_method ?
```

File: pymarsys/emarsys.py
Comment (misterjoa) [inline pymarsys/emarsys.py]: this part can be factorized for both sync and async code

**Comment by misterjoa** [comment]
have a look at this : http://stackoverflow.com/questions/30155138/how-can-i-write-asyncio-coroutines-that-optionally-act-as-regular-functions
it uses a metaclass template to produce sync version of async func

File: pymarsys/emarsys.py
Comment (Perados) [reply pymarsys/emarsys.py]: Indeed, but the whole settings thing for the constructor is not normal, after talking to @GaelMagnan , we decided to remove it.

File: pymarsys/emarsys.py
Comment (Perados) [reply pymarsys/emarsys.py]: I am not sure, the only examples I know use `get_event_loop`, but I will take a look.

File: pymarsys/emarsys.py
Comment (Perados) [reply pymarsys/emarsys.py]: Agreed.

File: pymarsys/emarsys.py
Comment (Perados) [reply pymarsys/emarsys.py]: Agreed. Will change.

**Comment by Perados** [comment]
@misterjoa After discussion with @GaelMagnan , the architecture will change a little bit. Using metaclasses is an option, but the result will probably be the same as it is now. I will take a look at many options and pick the most maintainable one.

**Comment by mknecht** [comment]
@Perados so, I'm supposed to review a PR whose architecture is about to change, and whose checks are failing? Try again. :p

File: setup.py
Comment (mknecht) [inline setup.py]: You're using async code? Do you require Python > 3.5?

File: requirements.txt
Comment (mknecht) [inline requirements.txt]: I'd reconsider the version pinning. I see a world of pain, filled with issues and forks.

Assume for a sec Appo, our app developer, takes care of his dependencies, and pins them. Then he tries to install your library, and *boom* he can't, because for whatever reason he uses requests `2.12.1`, but 2.13. Or maybe he follows the [two-files approach from Kenneth](https://www.kennethreitz.org/essays/a-better-pip-workflow) or the `https://github.com/pypa/pipfile` to pin indirect dependencies, too, and a library that they're already using needs `2.12.2`. Can't use your lib.

Remember, you're building a library, not an app. Any assumption you're making, you're forcing onto your users.

Comment (mknecht) [reply requirements.txt]: Follow-up: For your *tests*, to make them reproducible and explicit you may want to pin everything quite rigorously to known versions.

**Review by mknecht — CHANGES_REQUESTED** [review]
I'll look at the code, when the architecture is stable. :)

File: README.md
Comment (GaelMagnan) [inline README.md]: you should put an example with async/await.

File: pymarsys/connections.py
Comment (GaelMagnan) [inline pymarsys/connections.py]: thoses are http_headers too. and since the include the other ones, they should be called http_header, and the other called wsse_headers probably.

File: pymarsys/connections.py
Comment (GaelMagnan) [inline pymarsys/connections.py]: please keep the order between the doc and the param. + you describe a param called params but it doesn't exists.

File: pymarsys/connections.py
Comment (GaelMagnan) [inline pymarsys/connections.py]: i think using raise_for_status is better than testing the ok.

File: pymarsys/connections.py
Comment (GaelMagnan) [inline pymarsys/connections.py]: same comment as for synchronous.

File: pymarsys/connections.py
Comment (GaelMagnan) [inline pymarsys/connections.py]: You probably shouldn't create a session for each request, you can probably instantiate the session when you initialize the connection and use it for future calls.

File: pymarsys/connections.py
Comment (GaelMagnan) [inline pymarsys/connections.py]: same comment about raise_for_status.

File: pymarsys/contact.py
Comment (GaelMagnan) [inline pymarsys/contact.py]: this is not pure, we should probably make a copy of the object to ensure no modification happens to the data. And if not, at least write it in the doc.

File: requirements.txt
Comment (GaelMagnan) [reply requirements.txt]: The question of pinning or no pinning is a big debate, but i'd recommend creating a requirements.in file, specifying un-pinned dependencies, and one requirements.txt with the pinned ones. That way anyone can choose, and you can guarantee a version that works and has been tested.

File: README.md
Comment (Perados) [reply README.md]: Done.

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Indeed. Done, thanks!

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Fixed.

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Very useful indeed, thanks!

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Done.

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Done.

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Done.

File: pymarsys/contact.py
Comment (Perados) [reply pymarsys/contact.py]: :man_facepalming: 
Thanks!

**Comment by Perados** [comment]
@misterjoa @GaelMagnan @mknecht The pinning stuff is complicated. For now, I unpinned requests and aiohttp in the setup.py file. This will evolve over time, but I think it is ok for now.
I think the rest is ready for merging, waiting for your approval @misterjoa and @GaelMagnan . 😸

File: pymarsys/connections.py
Comment (misterjoa) [inline pymarsys/connections.py]: I think using this function will lead to wrong http adresses on windows systems.
use
```
'/'.join([...])
```
instead

File: pymarsys/connections.py
Comment (misterjoa) [inline pymarsys/connections.py:32]: [personal taste]: i would rather define this function as a staticmethod because it is not especially dependent of many class attributes, and using a staticmethod forces oneself to define function parameters

File: pymarsys/connections.py
Comment (misterjoa) [inline pymarsys/connections.py:48]: [personal taste]: i would rather define this function as a staticmethod because it is not especially dependent of many class attributes, and using a staticmethod forces oneself to define function parameters

File: pymarsys/connections.py
Comment (misterjoa) [inline pymarsys/connections.py]: so you rebuild the secret at each api call ?

File: pymarsys/connections.py
Comment (misterjoa) [inline pymarsys/connections.py]: same remark

File: pymarsys/contact.py
Comment (misterjoa) [inline pymarsys/contact.py:24]: if we aim to help the user, it would be nice to find a way, either to give the library user hints about standard fields, or do an input validation on field names.

**Review by misterjoa — CHANGES_REQUESTED** [review]

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py:32]: I don' t see a case where you would like to use this method to build authentication variables for a different connection, that would be weird. Let's keep it like this for now, this might change over time.

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py:48]: Same as above.

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: I was thinking about this the other day, but completely forgot to fix it. The right way to do it is to use urljoin from [urllib.parse](https://docs.python.org/3.0/library/urllib.parse.html). Thanks!

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Yes, each call needs to have a unique signature, that's what happens when you don't use oauth...

File: pymarsys/connections.py
Comment (Perados) [reply pymarsys/connections.py]: Done.

File: pymarsys/contact.py
Comment (Perados) [reply pymarsys/contact.py:24]: Will think about this for future prs.

**Review by misterjoa — APPROVED** [review]

**Pull request merged**
